op-batcher/batcher/channel_manager.go

```diff
@@ -343,19 +349,29 @@ func (s *channelManager) Close() error {
 	}
 
 	s.closed = true
+	s.log.Info("Channel manager is closing")
 
 	// Any pending state can be proactively cleared if there are no submitted transactions
 	for _, ch := range s.channelQueue {
 		if ch.NoneSubmitted() {
+			s.log.Info("Channel has no past or pending submission and might as well not exist, channel will be dropped", "channel", ch.ID(), "")
 			s.removePendingChannel(ch)
+		} else {
+			s.log.Info("Channel is in-flight and will need to be submitted after close", "channel", ch.ID(), "confirmed", len(ch.confirmedTransactions), "pending", len(ch.pendingTransactions))
 		}
 	}
+	s.log.Info("reviewed all pending channels on close", "remaining", len(s.channelQueue))
 
 	if s.currentChannel == nil {
 		return nil
 	}
 
+	// Force-close the remaining open channel early (if not already closed):
+	// it will be marked as "full" due to service termination.
 	s.currentChannel.Close()
 
-	return s.outputFrames()
+	// We do not s.outputFrames():
+	// this will only error if we cannot submit it in one go due to channel limits.
+	// Instead, make it clear to the caller that there is remaining pending work.
+	return ErrPendingAfterClose
```

Comment:
Why can we safely skip the `outputFrames()` call here? Where are the frames output instead? Note that the `TxData` call [exits early for closed channel managers](https://github.com/ethereum-optimism/optimism/blob/8a3c8964700f7890d021c76841c7853b1c4d5a89/op-batcher/batcher/channel_manager.go#L158) and doesn't call it.

I'm also wondering if we're missing a test (not the channel manager unit test) for this scenario, where the batcher is closed with a pending channel and correct submission is asserted.

Reply:
My impression was that the `outputFrames` call [here](https://github.com/ethereum-optimism/optimism/blob/8a3c8964700f7890d021c76841c7853b1c4d5a89/op-batcher/batcher/channel_manager.go#L183C14-L183C26) is sufficient. Is it possible that without the `outputFrames` here, it potentially submits less frame data in the first tx-data, but then still submits everything in the following txs? Note that `TestChannelManagerClosePendingChannel` passes.

Reply:
This `outputFrames` call should never be reached if called on a closed channel manager because of the early exit I linked in my first comment.

I'm worried that the test is no good, and that with `outputFrames` skipped, we might end up with a truncated channel. When the underlying `channel` is `Close`d, it has to be guaranteed that `OutputFrames` is called on it once, or the compressor is not flushed, afaik.

It's probably a good idea to expand the channel manager tests, at least this one, to assert that all the tx data output by the `TxData` calls, if reassembled, contains all the blocks that were previously put into the channel manager. I'm otherwise also not sure what's going on here, but I don't trust the test as it is.

Reply:
I'm not familiar with the batcher code so take this with a grain of salt, but I have the sense that there's a critical fork in `channelBuilder.Close` where it checks if the the channel is already marked full.  I suspect we want tests that cover both when it is and isn't full. `TestChannelManagerClosePendingChannel` is hitting the case where it's already full because `compressor is full`. I _think_ the net result of that is that all the data has already been output to frames and so it doesn't matter if we have this call to `outputFrames` or not.

With the channel already full, the call to `channelBuilder.OutputFrames` that results from the first call to `m.TxData` in the test will have already run the IsFull case where it calls `closeAndOutputAllFrames`.

I can't work out how to adjust the test to hit the other case where the full err actually gets set to `ErrTerminated` but I suspect we need to get there for this extra call to be important. Then `channelBuilder.OutputFrames` will only call `outputReadyFrames` initially and there will be some extra data that hasn't yet been added to the output frame list which would leave the channel unclosed.

It's telling that I can modify `channelBuilder.Close` to:
```
// Close immediately marks the channel as full with an ErrTerminated
// if the channel is not already full.
func (c *channelBuilder) Close() {
	println("Closing channel")
	if !c.IsFull() {
		panic("Setting ErrTerminated")
		c.setFullErr(ErrTerminated)
	}
}
```

and all the `channel_manager_test` cases still pass, so none of them is actually hitting the case where `channelBuilder.Close` is actually doing anything.

Reply:
Finally got around to this. Managed to write a test that does hit this code path and that shows that the call to `outputFrames` is actually needed.